azurerm/internal/services/storage/resource_arm_hpc_cache.go

```diff
@@ -0,0 +1,202 @@
+					6144,
+					12288,
+					24576,
+					49152,
+				}),
+			},
+
+			"subnet_id": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: azure.ValidateResourceIDOrEmpty,
+			},
+
+			"sku_name": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+				ValidateFunc: validation.StringInSlice([]string{
+					"Standard_2G",
+					"Standard_4G",
+					"Standard_8G",
+				}, false),
+			},
+		},
+	}
+}
+
+func resourceArmHPCCacheCreate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*clients.Client).Storage.CachesClient
+	ctx, cancel := timeouts.ForCreate(meta.(*clients.Client).StopContext, d)
+	defer cancel()
+
+	log.Printf("[INFO] preparing arguments for Azure HPC Cache creation.")
+	name := d.Get("name").(string)
+	resourceGroup := d.Get("resource_group_name").(string)
+
+	if features.ShouldResourcesBeImported() && d.IsNewResource() {
+		existing, err := client.Get(ctx, resourceGroup, name)
+		if err != nil {
+			if !utils.ResponseWasNotFound(existing.Response) {
+				return fmt.Errorf("Error checking for presence of existing HPC Cache %q (Resource Group %q): %s", name, resourceGroup, err)
+			}
+		}
+
+		if existing.ID != nil && *existing.ID != "" {
+			return tf.ImportAsExistsError("azurerm_hpc_cache", *existing.ID)
+		}
+	}
+
+	location := d.Get("location").(string)
+	cacheSize := d.Get("cache_size").(int)
+	subnet := d.Get("subnet_id").(string)
+	skuName := d.Get("sku_name").(string)
+
+	cache := &storagecache.Cache{
+		Name:     utils.String(name),
+		Location: utils.String(location),
+		CacheProperties: &storagecache.CacheProperties{
+			CacheSizeGB: utils.Int32(int32(cacheSize)),
+			Subnet:      utils.String(subnet),
+		},
+		Sku: &storagecache.CacheSku{
+			Name: utils.String(skuName),
+		},
+	}
+
+	future, err := client.CreateOrUpdate(ctx, resourceGroup, name, cache)
+	if err != nil {
+		return fmt.Errorf("Error creating HPC Cache %q (Resource Group %q): %+v", name, resourceGroup, err)
+	}
+
+	if err = future.WaitForCompletionRef(ctx, client.Client); err != nil {
+		return fmt.Errorf("Error waiting for HPC Cache %q (Resource Group %q) to finish provisioning: %+v", name, resourceGroup, err)
+	}
+
+	resp, err := client.Get(ctx, resourceGroup, name)
+	if err != nil {
+		return fmt.Errorf("Error retrieving HPC Cache %q (Resource Group %q): %+v", name, resourceGroup, err)
+	}
+	if resp.ID == nil {
+		return fmt.Errorf("Cannot read ID for HPC Cache %q (Resource Group %q): %+v", name, resourceGroup, err)
+	}
+	d.SetId(*resp.ID)
+
+	return resourceArmHPCCacheRead(d, meta)
+}
+
+func resourceArmHPCCacheRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*clients.Client).Storage.CachesClient
+	ctx, cancel := timeouts.ForRead(meta.(*clients.Client).StopContext, d)
+	defer cancel()
+
+	id, err := azure.ParseAzureResourceID(d.Id())
+	if err != nil {
+		return err
+	}
+	resourceGroup := id.ResourceGroup
+	name := id.Path["caches"]
```

Comment:
We are starting to write Parse ID functions for all resources, you can see an example here: https://github.com/terraform-providers/terraform-provider-azurerm/pull/5779